File: ts/medium/myFlat.ts
type NestedNumberArray = number | NestedNumberArray[];

const a: NestedNumberArray = [1, 2, 3, [4, 5, 6]];

function myFlat(arr: NestedNumberArray): number[] {
  const acc: number[] = [];

  if (Array.isArray(arr)) {
    for (const item of arr) {
      if (Array.isArray(item)) {
        acc.push(...myFlat(item));
      } else {
        acc.push(item);
      }
    }
  } else {
    acc.push(arr);
  }

  return acc;
}

const flattened = myFlat(a);
console.log(flattened);

File: python/medium/my_flat.py
from typing import List, Union

NestedNumberList = Union[int, List["NestedNumberList"]]

l = [1,2,3, [4,5,6]]

def my_flat(arr: NestedNumberList) -> List[int]:
    acc = []
    
    if isinstance(arr, list):
        for item in arr:
            acc.extend(my_flat(item))
    else:
        acc.append(arr)
    return acc
print(my_flat(l))
